server/storage/bootstrap.go

```diff
@@ -0,0 +1,335 @@
+	BeExist  bool
+	Snapshot *raftpb.Snapshot
+}
+
+type BootstrappedWAL struct {
+	lg *zap.Logger
+
+	HaveWAL  bool
+	Wal      *wal.WAL
+	State    *raftpb.HardState
+	Entries  []raftpb.Entry
+	Snapshot *raftpb.Snapshot
+	Meta     *Metadata
+}
+
+type Metadata struct {
+	NodeID, ClusterID types.ID
+}
+
+func BootstrapStorage(cfg config.ServerConfig, st v2store.Store, be *bootstrappedBackend, wal *BootstrappedWAL, meta *Metadata) (b *BootstrappedStorage, err error) {
+	if wal == nil {
+		wal = bootstrapNewWAL(cfg, meta)
+	}
+
+	return &BootstrappedStorage{
+		Backend: be,
+		St:      st,
+		Wal:     wal,
+	}, nil
+}
+
+func BootstrapBackend(cfg config.ServerConfig, st v2store.Store, ss *snap.Snapshotter) (backend *bootstrappedBackend, err error) {
+	beExist := fileutil.Exist(cfg.BackendPath())
+	ci := cindex.NewConsistentIndex(nil)
+	beHooks := NewBackendHooks(cfg.Logger, ci)
+	be := OpenBackend(cfg, beHooks)
+	defer func() {
+		if err != nil {
+			be.Close()
+		}
+	}()
+	ci.SetBackend(be)
+	schema.CreateMetaBucket(be.BatchTx())
+	if cfg.ExperimentalBootstrapDefragThresholdMegabytes != 0 {
+		err = maybeDefragBackend(cfg, be)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cfg.Logger.Debug("restore consistentIndex", zap.Uint64("index", ci.ConsistentIndex()))
+	var (
+		walSnaps []walpb.Snapshot
+		snapshot *raftpb.Snapshot
+	)
+
+	if beExist {
+		if err = fileutil.IsDirWriteable(cfg.WALDir()); err != nil {
+			return nil, fmt.Errorf("cannot write to WAL directory: %v", err)
+		}
+
+		// Find a snapshot to start/restart a raft node
+		walSnaps, err = wal.ValidSnapshotEntries(cfg.Logger, cfg.WALDir())
+		if err != nil {
+			return nil, err
+		}
+		// Snapshot files can Be orphaned if etcd crashes after writing them but before writing the corresponding
+		// Wal log entries
+		snapshot, err = ss.LoadNewestAvailable(walSnaps)
+		if err != nil && err != snap.ErrNoSnapshot {
+			return nil, err
+		}
+
+		if snapshot != nil {
+			if err = st.Recovery(snapshot.Data); err != nil {
+				cfg.Logger.Panic("failed to recover from snapshot", zap.Error(err))
+			}
+
+			if err = AssertNoV2StoreContent(cfg.Logger, st, cfg.V2Deprecation); err != nil {
+				cfg.Logger.Error("illegal v2store content", zap.Error(err))
+				return nil, err
+			}
+
+			cfg.Logger.Info(
+				"recovered v2 store from snapshot",
+				zap.Uint64("snapshot-index", snapshot.Metadata.Index),
+				zap.String("snapshot-size", humanize.Bytes(uint64(snapshot.Size()))),
+			)
+
+			if be, err = RecoverSnapshotBackend(cfg, be, *snapshot, beExist, beHooks); err != nil {
+				cfg.Logger.Panic("failed to recover v3 backend from snapshot", zap.Error(err))
+			}
+			s1, s2 := be.Size(), be.SizeInUse()
+			cfg.Logger.Info(
+				"recovered v3 backend from snapshot",
+				zap.Int64("backend-size-bytes", s1),
+				zap.String("backend-size", humanize.Bytes(uint64(s1))),
+				zap.Int64("backend-size-in-use-bytes", s2),
+				zap.String("backend-size-in-use", humanize.Bytes(uint64(s2))),
+			)
+			if beExist {
```

Comment:
We are already in the beExist branch.  Do we need the second check ?

Reply:
Done